File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min_to_buy = prices[0];
        int max_profit = 0;
        int cur_profit;
        for(int i=1; i<prices.size(); i++){
            if(prices[i]-min_to_buy<=0) cur_profit = 0;
            else cur_profit = prices[i]-min_to_buy;
            max_profit = max(max_profit, cur_profit);
            min_to_buy = min(min_to_buy, prices[i]);
        }
        return max_profit;
    }
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution(object):
    def maxProfit(self, prices):
        """
        :type prices: List[int]
        :rtype: int
        """
        low = 0
        high = 0
        lowest = prices[0]
        highest = prices[0]
        profit = highest - lowest
        while high < len(prices) -1 and low < len(prices) - 1:
            if prices[high + 1] > highest:
                high += 1
                highest = prices[high]
                if highest - lowest > profit:
                    profit = highest - lowest
            elif prices[high + 1] < lowest:
                low = high + 1
                high = high + 1
                lowest = prices[low]
                highest = prices[high]
            else:
                high += 1
        return profit
        